src/wp-includes/html-api/class-wp-html-tag-processor.php

```diff
@@ -1431,8 +1431,15 @@ private function skip_script_data(): bool {
 				continue;
 			}
 
-			// Everything of interest past here starts with "<".
-			if ( $at + 1 >= $doc_length || '<' !== $html[ $at++ ] ) {
+			if ( $at + 1 >= $doc_length ) {
```

Comment:
I've chosen to break here because there's no way to find the script closer. The document is incomplete, so breaking and returning `false` below enters the correct code path. This could also `return false` directly instead of breaking out of the loop.

This condition could likely be increased. We need to match at least `<!--` so we could apply this with at least `$at + 3`.

```suggestion
			// The parser needs to match at least `<!--` from here.
			if ( $at + 3 >= $doc_length ) {
```

---

I also think this could apply more aggressive length checks earlier, then skip all of the subsequent length checks. Ultimately, we need to find a script closer in order to exit `true`. It doesn't matter how much of the script contents are parsed if it can't be closed the parser will stop on incomplete input. After this line:

https://github.com/WordPress/wordpress-develop/blob/a799101e46ef0a042814c189a6331c23a52a100c/src/wp-includes/html-api/class-wp-html-tag-processor.php#L1416

We could include this to eagerly reach incomplete input if we know a closer cannot be found.

```php
if ( $at + 8 > $doc_length ) { return false; }
```

Here's my thinking:

```php
/* Things could happen in the loop, but they're irrelevant if a `</script>` cannot be found  */

// $at:  V
<script> ----->␃
// $at+8         ^

/* A script closer could fit here and is ultimately found */

// $at:          V
<script> bla bla </script>␃
// $at+8                 ^

/* A script closer would fit so processing continues */
// $at:          V
<script> bla bla -            more document…␃
// $at+8                 ^
```

Maybe the proposed bugfix could be landed and included in WordPress 6.6, with more optimizations included landed in another change to `trunk`.

@dmsnell curious to hear your thoughts.

Reply:
these are valid points. what I like about advancing one character at a time is that in some cases, I leaned on that to ensure we didn't count wrong. a targeting fix sounds nice now, and then a follow-up can be tested in isolation, particularly for performance, for skipping ahead.

Reply:
I think I also like returning `false` directly - no need to further confuse control flow